server/api/src/pb/data_access.py

```diff
@@ -6,64 +6,87 @@
 
 
 DATA_DIR = os.path.join(Server.TMP_DIR, 'picklebase')
+STAGE_DIR = os.path.join(Server.TMP_DIR, 'picklebase-stage')
 READY_FILE = os.path.join(DATA_DIR, 'ready')
 
 
+# ########## RESETS ########### #
+
+def clear_stage():
+    shutil.rmtree(STAGE_DIR, ignore_errors=True)
+
+
 def clear_data():
     shutil.rmtree(DATA_DIR, ignore_errors=True)
 
 
+# ###### PATHS / FILENAMES ##### #
+
+def stage_table_path(table):
+    return os.path.join(STAGE_DIR, table)
+
+
 def table_path(table):
     return os.path.join(DATA_DIR, table)
 
 
-def init_table(table):
-    path = table_path(table)
-    shutil.rmtree(path, ignore_errors=True)
-    os.makedirs(path, exist_ok=True)
+def stage_meta_path(table):
+    return os.path.join(STAGE_DIR, table, 'meta.json')
+
+
+def meta_path(table):
+    return os.path.join(DATA_DIR, table, 'meta.json')
 
 
 def batch_filename(batch_num):
     return f'batch_{batch_num}'
 
 
+# ############ WRITING ########### #
+
+def init_table(table):
+    path = stage_table_path(table)
+    shutil.rmtree(path, ignore_errors=True)
+    os.makedirs(path, exist_ok=True)
+
+
 def save_batch(table, batch_num, batch):
     filename = batch_filename(batch_num)
-    path = os.path.join(table_path(table), filename)
+    path = os.path.join(stage_table_path(table), filename)
     with open(path, 'wb') as f:
         pickle.dump(batch, f)
     return filename, os.path.getsize(path)
 
 
-def load_batch(table, batch_num):
-    path = os.path.join(table_path(table), batch_filename(batch_num))
-    with open(path, 'rb') as f:
-        return pickle.load(f)
-
-
-def meta_path(table):
-    return os.path.join(DATA_DIR, table, 'meta.json')
-
-
 def save_meta(table, meta):
-    path = meta_path(table)
+    path = stage_meta_path(table)
     meta_json = json.dumps(meta, indent=2)
     print('\nSaving meta:', flush=True)
     print(meta_json, flush=True)
     with open(path, 'w') as f:
         f.write(meta_json)
 
 
+def commit_pb():
+    shutil.rmtree(DATA_DIR, ignore_errors=True)
+    shutil.move(STAGE_DIR, DATA_DIR)
+    with open(READY_FILE, 'w'):
+        pass
```

Comment:
What's the purpose of this?

Reply:
Good question, that line just creates a file called 'ready', which lets the picklebase client(s) know that the picklebase has been populated and is ready to be queried. It's only necessary for heroku, where we have to populate the picklebase concurrently with starting the server because of the fact that heroku deletes the filesystem whenever the dyno restarts. 

The reason to use a file (as opposed to setting a boolean variable) is that Sanic runs in multiple workers, and the workers don't share memory. But they do share the filesystem. So creating a file is essentially like setting a boolean variable that's available to all of the workers.